include/SDL3/SDL_ioasync.h

```diff
@@ -0,0 +1,500 @@
+ * \param ptr a pointer to a buffer to read data into.
+ * \param offset the position to start reading in the data source.
+ * \param size the number of bytes to read from the data source.
+ * \param group a group to add the new SDL_IOAsync to.
+ * \param userdata an app-defined pointer that will be provided with the task results.
+ * \returns A new task handle if a task was started, NULL on complete failure;
+ *          call SDL_GetError() for more information.
+ *
+ * \threadsafety It is safe to call this function from any thread.
+ *
+ * \since This function is available since SDL 3.0.0.
+ *
+ * \sa SDL_WriteIOAsync
+ * \sa SDL_CreateIOAsyncGroup
+ * \sa SDL_GetIOAsyncTaskResult
+ */
+extern SDL_DECLSPEC SDL_IOAsyncTask * SDLCALL SDL_ReadIOAsync(SDL_IOAsync *ioasync, void *ptr, Uint64 offset, Uint64 size, SDL_IOAsyncGroup *group, void *userdata);
+
+/**
+ * Start an async write.
+ *
+ * This function writes `size` bytes from `offset` position in the data
+ * source to the area pointed at by `ptr`.
+ *
+ * This function returns as quickly as possible; it does not wait for the
+ * write to complete. On a successful return, this work will continue in the
+ * background. If the work begins, even failure is asynchronous: a failing
+ * return value from this function only means the work couldn't start at all.
+ *
+ * `ptr` must remain available until the work is done, and may be accessed by
+ * the system at any time until then. Do not allocate it on the stack, as this
+ * might take longer than the life of the calling function to complete!
+ *
+ * An SDL_IOAsyncGroup must be specified. The newly-created SDL_IOAsyncTask
+ * will be added to it. A group is used to query multiple async tasks at once.
+ *
+ * \param ioasync a pointer to an SDL_IOAsync structure.
+ * \param ptr a pointer to a buffer to write data from.
+ * \param offset the position to start writing to the data source.
+ * \param size the number of bytes to write to the data source.
+ * \param group a group to add the new SDL_IOAsync to.
+ * \param userdata an app-defined pointer that will be provided with the task results.
+ * \returns A new task handle if a task was started, NULL on complete failure;
+ *          call SDL_GetError() for more information.
+ *
+ * \threadsafety It is safe to call this function from any thread.
+ *
+ * \since This function is available since SDL 3.0.0.
+ *
+ * \sa SDL_ReadIOAsync
+ * \sa SDL_CreateIOAsyncGroup
+ * \sa SDL_GetIOAsyncTaskResult
+
+ */
+extern SDL_DECLSPEC SDL_IOAsyncTask * SDLCALL SDL_WriteIOAsync(SDL_IOAsync *ioasync, void *ptr, Uint64 offset, Uint64 size, SDL_IOAsyncGroup *group, void *userdata);
+
+/**
+ * Close and free any allocated resources for an async I/O object.
+ *
+ * Closing a file is _also_ an asynchronous task! If a write failure
+ * were to happen during the closing process, for example, the
+ * task results will report it as usual.
+ *
+ * This function guarantees that the close will happen after any other
```

Comment:
Does closing guarantee a flush, or could data be left in the page cache? Maybe for games it doesn't matter, but flushing is required to build a save strategy that doesn't corrupt on power failure or OS crash.

Reply:
I'm assuming it implies the same flush that close/fclose/CloseHandle would (which is literally what the generic backend does, but I assume this would be true for IoRing/io_uring, too).

Reply:
From `man close` on POSIX:

>   A successful close does not guarantee that the data has been successfully saved to disk, as the kernel uses the buffer cache  to defer writes. Typically, filesystems do not flush buffers when a file is closed.  If you need to be sure that the data is physically stored on the underlying disk, use [fsync(2)](https://man7.org/linux/man-pages/man2/fsync.2.html).

You need an asynchronous fsync baked into this close operation for this to really be correct. I think the comment here should say that this operation guarantees all data has been flushed from the operating system caches before it completes - we'll save a lot of developers a lot of headaches around corrupt files during OS crash or power loss.

Reply:
the flush that fclose/fflush does just flushes the libc stream buffers by calling write/WriteFile, it doesn't ensure that writes buffered by the OS are committed to disk - the C standard really has no concept of OS write buffering.

close and CloseHandle won't ensure that either, you need to explicitly use fsync/FlushFileBuffers if you want this.

Reply:
We should not add fsync to this API. If we want that, it should be something separate that we add for iostream and asyncio

Reply:
I just want as good as possible sync-on-close semantics and I believe that's all jaedan wants too.

Reply:
Yeah for SDL purposes the common use cases are writing config/settings/profiles, and writing out save state. For both of those operations, you need to do the following pattern:

1) asynchronously write out the file data to a temp file
2) asynchronously flush the OS page cache for that file. You don't want to block here, it can take a long time.
3) delete the old file
4) rename the new file to the correct name

Step 2 can be baked into closing the file for simplicity. And step 3 and 4 are usually fast and do not need to be asynchronous (although I won't argue against making it all async).

This whole process doesn't have to necessarily be captured in this API but correct, cross-platform, asynchronous sync-on-close would be extremely valuable. Very few game devs end up getting this right on their own.

Reply:
To be pedantic, step 3 shouldn't be done, since the rename is atomic by itself; you end up with either the old data at that name or the new data but never halfway or both.

Which also makes me wonder in this pattern how important fsync actually is, since you'll either have the old data undamaged or the new data correctly in its place, flushing the file cache maybe isn't that important for data integrity here.

Reply:
But you do need the flush between the writes and the rename because if you don't, you could crash after the rename and be left with a corrupt file.

Reply:
yes, rename is atomic WRT the inodes but not WRT file contents. fsync before rename is essential.